File: 98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.DFS_helper(root, float('-inf'), float('inf'))
        
    def DFS_helper(self, root, min, max):
        if not root:
            return True
        if root.val >= max or root.val <= min:
            return False
        return self.DFS_helper(root.left, min, root.val) and self.DFS_helper(root.right, root.val, max)

File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (!root)
            return true;
        return BST(root, LONG_MIN, LONG_MAX);
    }
    bool BST(TreeNode* root, long min, long max){
        if(!root)
            return true;
        if(root->val >= max || root->val <= min)
            return false;
        return BST(root->left, min, root->val) && BST(root->right, root->val, max);
    }

};
